File: gnd/src/main/java/com/google/android/gnd/MainViewModel.java
/*
 * Copyright 2018 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.google.android.gnd;

import static com.google.android.gnd.rx.RxTransformers.switchMapIfPresent;

import androidx.core.view.WindowInsetsCompat;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.navigation.NavDirections;
import com.google.android.gnd.model.Project;
import com.google.android.gnd.model.TermsOfService;
import com.google.android.gnd.repository.FeatureRepository;
import com.google.android.gnd.repository.ProjectRepository;
import com.google.android.gnd.repository.TermsOfServiceRepository;
import com.google.android.gnd.repository.UserRepository;
import com.google.android.gnd.rx.Schedulers;
import com.google.android.gnd.rx.annotations.Cold;
import com.google.android.gnd.rx.annotations.Hot;
import com.google.android.gnd.system.auth.AuthenticationManager;
import com.google.android.gnd.system.auth.SignInState;
import com.google.android.gnd.system.auth.SignInState.State;
import com.google.android.gnd.ui.common.AbstractViewModel;
import com.google.android.gnd.ui.common.EphemeralPopups;
import com.google.android.gnd.ui.common.Navigator;
import com.google.android.gnd.ui.common.SharedViewModel;
import com.google.android.gnd.ui.home.HomeScreenFragmentDirections;
import com.google.android.gnd.ui.signin.SignInFragmentDirections;
import io.reactivex.Completable;
import io.reactivex.Maybe;
import io.reactivex.Observable;
import java8.util.Optional;
import javax.inject.Inject;
import timber.log.Timber;

/** Top-level view model representing state of the {@link MainActivity} shared by all fragments. */
@SharedViewModel
public class MainViewModel extends AbstractViewModel {

  /** The window insets determined by the activity. */
  @Hot(replays = true)
  private final MutableLiveData<WindowInsetsCompat> windowInsets = new MutableLiveData<>();

  /** The state of sign in progress dialog visibility. */
  @Hot(replays = true)
  private final MutableLiveData<Boolean> signInProgressDialogVisibility = new MutableLiveData<>();

  private final ProjectRepository projectRepository;
  private final FeatureRepository featureRepository;
  private final UserRepository userRepository;
  private final TermsOfServiceRepository termsOfServiceRepository;
  private final EphemeralPopups popups;

  @Inject
  public MainViewModel(
      ProjectRepository projectRepository,
      FeatureRepository featureRepository,
      UserRepository userRepository,
      TermsOfServiceRepository termsOfServiceRepository,
      Navigator navigator,
      AuthenticationManager authenticationManager,
      EphemeralPopups popups,
      Schedulers schedulers) {
    this.projectRepository = projectRepository;
    this.featureRepository = featureRepository;
    this.termsOfServiceRepository = termsOfServiceRepository;
    this.userRepository = userRepository;
    this.popups = popups;

    // TODO: Move to background service.
    disposeOnClear(
        projectRepository
            .getActiveProject()
            .observeOn(schedulers.io())
            .switchMapCompletable(this::syncFeatures)
            .subscribe());

    disposeOnClear(
        authenticationManager
            .getSignInState()
            .compose(switchMapIfPresent(SignInState::getUser, userRepository::saveUser))
            .observeOn(schedulers.ui())
            .switchMap(this::onSignInStateChange)
            .subscribe(navigator::navigate));
  }

  /**
   * Keeps local features in sync with remote when a project is active, does nothing when no project
   * is active. The stream never completes; syncing stops when subscriptions are disposed of.
   *
   * @param project the currently active project.
   */
  @Cold(terminates = false)
  private Completable syncFeatures(Optional<Project> project) {
    return project.map(featureRepository::syncFeatures).orElse(Completable.never());
  }

  @Hot(replays = true)
  public LiveData<WindowInsetsCompat> getWindowInsets() {
    return windowInsets;
  }

  void onApplyWindowInsets(WindowInsetsCompat insets) {
    windowInsets.setValue(insets);
  }

  private Observable<NavDirections> onSignInStateChange(SignInState signInState) {
    if (signInState.state() != State.SIGNED_IN) {
      termsOfServiceRepository.setTermsOfServiceAccepted(false);
      featureRepository.setPolygonDialogInfoShown(false);
    }

    switch (signInState.state()) {
      case SIGNED_OUT:
        // TODO: Check auth status whenever fragments resumes.
        return onSignedOut();
      case SIGNING_IN:
        showProgressDialog();
        break;
      case SIGNED_IN:
        return onSignedIn();
      case ERROR:
        Timber.d("Authentication error : %s", signInState.error());
        return onSignInError();
      default:
        Timber.e("Unhandled state: %s", signInState.state());
        break;
    }
    return Observable.never();
  }

  private void showProgressDialog() {
    signInProgressDialogVisibility.postValue(true);
  }

  private void hideProgressDialog() {
    signInProgressDialogVisibility.postValue(false);
  }

  private Observable<NavDirections> onSignInError() {
    popups.showError(R.string.sign_in_unsuccessful);
    return onSignedOut();
  }

  private Observable<NavDirections> onSignedOut() {
    hideProgressDialog();
    projectRepository.clearActiveProject();
    userRepository.clearUserPreferences();
    return Observable.just(SignInFragmentDirections.showSignInScreen());
  }

  private Observable<NavDirections> onSignedIn() {
    hideProgressDialog();
    if (termsOfServiceRepository.isTermsOfServiceAccepted()) {
      return Observable.just(HomeScreenFragmentDirections.showHomeScreen());
    } else {
      return termsOfServiceRepository
          .getTermsOfService()
          .map(TermsOfService::getText)
          .map(text -> SignInFragmentDirections.showTermsOfService().setTermsOfServiceText(text))
          .cast(NavDirections.class)
          .switchIfEmpty(Maybe.just(HomeScreenFragmentDirections.showHomeScreen()))
          .toObservable();
    }
  }

  public LiveData<Boolean> getSignInProgressDialogVisibility() {
    return signInProgressDialogVisibility;
  }
}

File: gnd/src/main/java/com/google/android/gnd/MainViewModel.kt
/*
 * Copyright 2018 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.google.android.gnd

import androidx.core.view.WindowInsetsCompat
import androidx.lifecycle.MutableLiveData
import androidx.navigation.NavDirections
import com.google.android.gnd.model.Survey
import com.google.android.gnd.repository.FeatureRepository
import com.google.android.gnd.repository.SurveyRepository
import com.google.android.gnd.repository.TermsOfServiceRepository
import com.google.android.gnd.repository.UserRepository
import com.google.android.gnd.rx.RxTransformers.switchMapIfPresent
import com.google.android.gnd.rx.Schedulers
import com.google.android.gnd.rx.annotations.Cold
import com.google.android.gnd.system.auth.AuthenticationManager
import com.google.android.gnd.system.auth.SignInState
import com.google.android.gnd.ui.common.AbstractViewModel
import com.google.android.gnd.ui.common.EphemeralPopups
import com.google.android.gnd.ui.common.Navigator
import com.google.android.gnd.ui.common.SharedViewModel
import com.google.android.gnd.ui.home.HomeScreenFragmentDirections
import com.google.android.gnd.ui.signin.SignInFragmentDirections
import io.reactivex.Completable
import io.reactivex.Maybe
import io.reactivex.Observable
import java8.util.Optional
import timber.log.Timber
import javax.inject.Inject

/** Top-level view model representing state of the [MainActivity] shared by all fragments.  */
@SharedViewModel
class MainViewModel @Inject constructor(
    private val surveyRepository: SurveyRepository,
    private val featureRepository: FeatureRepository,
    private val userRepository: UserRepository,
    private val termsOfServiceRepository: TermsOfServiceRepository,
    private val popups: EphemeralPopups,
    navigator: Navigator,
    authenticationManager: AuthenticationManager,
    schedulers: Schedulers
) : AbstractViewModel() {

    /** The window insets determined by the activity.  */
    val windowInsets: MutableLiveData<WindowInsetsCompat> = MutableLiveData()

    /** The state of sign in progress dialog visibility.  */
    val signInProgressDialogVisibility: MutableLiveData<Boolean> = MutableLiveData()

    init {
        // TODO: Move to background service.
        disposeOnClear(
            surveyRepository
                .activeSurvey
                .observeOn(schedulers.io())
                .switchMapCompletable { syncFeatures(it) }
                .subscribe()
        )

        disposeOnClear(
            authenticationManager
                .signInState
                .compose(switchMapIfPresent(SignInState::user) { userRepository.saveUser(it) })
                .observeOn(schedulers.ui())
                .switchMap { signInState: SignInState -> onSignInStateChange(signInState) }
                .subscribe { directions: NavDirections -> navigator.navigate(directions) })
    }

    /**
     * Keeps local features in sync with remote when a survey is active, does nothing when no survey
     * is active. The stream never completes; syncing stops when subscriptions are disposed of.
     *
     * @param survey the currently active survey.
     */
    private fun syncFeatures(survey: Optional<Survey>): @Cold(terminates = false) Completable {
        return survey.map { featureRepository.syncFeatures(it) }
            .orElse(Completable.never())
    }

    private fun onSignInStateChange(signInState: SignInState): Observable<NavDirections> {
        // Display progress only when signing in.
        signInProgressDialogVisibility.postValue(signInState.state == SignInState.State.SIGNING_IN)

        // TODO: Check auth status whenever fragments resumes
        return when (signInState.state) {
            SignInState.State.SIGNED_IN -> onUserSignedIn()
            SignInState.State.SIGNED_OUT -> onUserSignedOut()
            SignInState.State.ERROR -> onUserSignInError(signInState.error())
            else -> Observable.never()
        }
    }

    private fun onUserSignInError(error: Optional<Throwable?>): Observable<NavDirections> {
        Timber.e("Authentication error: $error")
        popups.showError(R.string.sign_in_unsuccessful)
        return onUserSignedOut()
    }

    private fun onUserSignedOut(): Observable<NavDirections> {
        surveyRepository.clearActiveSurvey()
        userRepository.clearUserPreferences()
        return Observable.just(SignInFragmentDirections.showSignInScreen())
    }

    private fun onUserSignedIn(): Observable<NavDirections> {
        return if (termsOfServiceRepository.isTermsOfServiceAccepted) {
            Observable.just(HomeScreenFragmentDirections.showHomeScreen())
        } else {
            termsOfServiceRepository
                .termsOfService
                .map {
                    SignInFragmentDirections.showTermsOfService().setTermsOfServiceText(it.text)
                }
                .cast(NavDirections::class.java)
                .switchIfEmpty(Maybe.just(HomeScreenFragmentDirections.showHomeScreen()))
                .toObservable()
        }
    }
}
